ingester/ingester_lifecycle.go

```diff
@@ -0,0 +1,419 @@
+			// then pick some tokens and enter ACTIVE state.
+			if i.state == ring.PENDING {
+				log.Infof("Auto-joining cluster after timout.")
+				if err := i.autoJoin(); err != nil {
+					log.Fatalf("failed to pick tokens in consul: %v", err)
+				}
+			}
+
+		case <-heartbeatTicker.C:
+			consulHeartbeats.Inc()
+			if err := i.updateConsul(); err != nil {
+				log.Errorf("Failed to write to consul, sleeping: %v", err)
+			}
+
+		case <-flushTicker.C:
+			i.sweepUsers(false)
+
+		case <-rateUpdateTicker.C:
+			i.userStates.updateRates()
+
+		case f := <-i.actorChan:
+			f()
+
+		case <-i.quit:
+			break loop
+		}
+	}
+
+	// Shutdown order is important!
+	i.changeState(ring.LEAVING)
+
+	flushRequired := true
+	if i.cfg.ClaimOnRollout {
+		if err := i.transferChunks(); err != nil {
+			log.Fatalf("Failed to transfer chunks to another ingester: %v", err)
+		}
+		flushRequired = false
+	}
+
+	if flushRequired {
+		i.flushAllChunks()
+	}
+
+	if !i.cfg.skipUnregister {
+		if err := i.unregister(); err != nil {
+			log.Fatalf("Failed to unregister from consul: %v", err)
+		}
+		log.Infof("Ingester removed from consul")
+	}
+
+	for _, flushQueue := range i.flushQueues {
+		flushQueue.Close()
+	}
+}
+
+// initRing is the first thing we do when we start. It:
+// - add an ingester entry to the ring
+// - copies out our state and tokens if they exist
+func (i *Ingester) initRing() error {
+	return i.consul.CAS(ring.ConsulKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		var ringDesc *ring.Desc
+		if in == nil {
+			ringDesc = ring.NewDesc()
+		} else {
+			ringDesc = in.(*ring.Desc)
+		}
+
+		ingesterDesc, ok := ringDesc.Ingesters[i.id]
+		if !ok {
+			// Either we are a new ingester, or consul must have restarted
+			log.Infof("Entry not found in ring, adding with no tokens.")
+			ringDesc.AddIngester(i.id, i.addr, []uint32{}, i.state)
+			return ringDesc, true, nil
+		}
+
+		// We exist in the ring, so assume the ring is right and copy out tokens & state out of there.
+		i.state = ingesterDesc.State
+		for _, token := range ringDesc.Tokens {
+			if token.Ingester == i.id {
+				i.tokens = append(i.tokens, token.Token)
+			}
+		}
+
+		log.Infof("Existing entry found in ring with state=%s, tokens=%v.", i.state, i.tokens)
+		return ringDesc, true, nil
+	})
+}
+
+// autoJoin selects random tokens & moves state to ACTIVE
+func (i *Ingester) autoJoin() error {
+	return i.consul.CAS(ring.ConsulKey, func(in interface{}) (out interface{}, retry bool, err error) {
+		var ringDesc *ring.Desc
+		if in == nil {
+			ringDesc = ring.NewDesc()
+		} else {
+			ringDesc = in.(*ring.Desc)
+		}
+
+		// At this point, we should not have any tokens, and we should be in PENDING state.
+
+		var takenTokens, myTokens []uint32
+		for _, token := range ringDesc.Tokens {
+			takenTokens = append(takenTokens, token.Token)
+			if token.Ingester == i.id {
+				myTokens = append(myTokens, token.Token)
+			}
+		}
```

Comment:
This has appeared a couple of times now. Worth factoring out a `filterByIngester` function?